tensorflow-framework/src/main/java/org/tensorflow/framework/data/DatasetIterator.java

```diff
@@ -0,0 +1,242 @@
+ *
+ *  int BATCH_SIZE = ...
+ *
+ *  Dataset dataset = Dataset
+ *          .fromTensorSlices(XTensor, yTensor)
+ *          .batch(BATCH_SIZE);
+ *  DatasetIterator iterator = dataset.makeIterator();
+ *
+ *  Optimizer optimizer = ... // create an optimizer
+ *
+ *  for (List<Output<?>> components : dataset) {
+ *      Operand<?> XBatch = components.get(0);
+ *      Operand<?> yBatch = components.get(1);
+ *
+ *      loss = computeModelLoss(X, y);
+ *      trainOp = optimizer.minimize(loss);
+ *  }
+ * }</pre>
+ */
+public class DatasetIterator {
+  public static final String EMPTY_SHARED_NAME = "";
+
+  private Ops tf;
+
+  private Operand<?> iteratorResource;
+  private Op initializer;
+
+  private List<DataType<?>> outputTypes;
+  private List<Shape> outputShapes;
+
+  /**
+   * @param tf               Ops accessor corresponding to the same `ExecutionEnvironment`
+   *                         as the `iteratorResource`.
+   * @param iteratorResource An Operand representing the iterator
+   *                         (e.g. constructed from `tf.data.iterator` or
+   *                         `tf.data.anonymousIterator`)
+   * @param initializer      An `Op` that should be run to initialize this iterator
+   * @param outputTypes      A list of `DataType` objects corresponding to the
+   *                         types of each component of a dataset element.
+   * @param outputShapes     A list of `Shape` objects corresponding to the
+   *                         shapes of each componenet of a dataset element.
+   */
+  private DatasetIterator(Ops tf, Operand<?> iteratorResource,
+                          Op initializer,
+                          List<DataType<?>> outputTypes,
+                          List<Shape> outputShapes) {
+
+    this.tf = tf;
+    this.iteratorResource = iteratorResource;
+    this.initializer = initializer;
+    this.outputTypes = outputTypes;
+    this.outputShapes = outputShapes;
+  }
+
+  private DatasetIterator(Ops tf, Operand<?> iteratorResource,
+                          List<DataType<?>> outputTypes,
+                          List<Shape> outputShapes) {
+    this.tf = tf;
+    this.iteratorResource = iteratorResource;
+    this.outputTypes = outputTypes;
+    this.outputShapes = outputShapes;
+  }
+
+  /**
+   * Returns a list of `Operand<?>` representing the components of the
+   * next dataset element.
+   * <p>
+   * In graph mode, call this method once, and use its result as input
+   * to another computation. Then in the training loop, on successive calls
+   * to session.run(), successive dataset elements will be retrieved through
+   * these components.
+   * <p>
+   * In eager mode, each time this method is called, the next dataset
+   * element will be returned. (This is done automatically by iterating
+   * through `Dataset` as a Java `Iterable`).
+   *
+   * @return A `List<Operand<?>>` representing dataset element components.
+   */
+  public List<Output<?>> getNext() {
```

Comment:
These should be `List<Output<?>>`. We could change them all to `List<Operand<?>>` if we modify the `IteratorGetNext` classes; since these are generated I wasn't sure how to do that.

Think that change would be worth it?

Reply:
These all have been changed to `List<Operand<?>>`